cub/test/catch2_test_device_reduce_nondeterministic.cu

```diff
@@ -0,0 +1,298 @@
+// SPDX-FileCopyrightText: Copyright (c) 2025, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+// SPDX-License-Identifier: BSD-3-Clause
+
+#include "insert_nested_NVTX_range_guard.h"
+
+#include <cub/device/dispatch/dispatch_reduce_nondeterministic.cuh>
+#include <cub/util_type.cuh>
+
+#include <thrust/device_vector.h>
+
+#include <cuda/__execution/determinism.h>
+#include <cuda/__execution/require.h>
+
+#include <numeric>
+
+#include "catch2_test_device_reduce.cuh"
+#include <c2h/catch2_test_helper.h>
+#include <c2h/generators.h>
+
+using float_type_list = c2h::type_list<float, double>;
+
+template <int NOMINAL_BLOCK_THREADS_4B, int NOMINAL_ITEMS_PER_THREAD_4B>
+struct AgentReducePolicy
+{
+  /// Number of items per vectorized load
+  static constexpr int VECTOR_LOAD_LENGTH = 4;
+
+  /// Cooperative block-wide reduction algorithm to use
+  static constexpr cub::BlockReduceAlgorithm BLOCK_ALGORITHM =
+    cub::BlockReduceAlgorithm::BLOCK_REDUCE_WARP_REDUCTIONS_NONDETERMINISTIC;
+
+  /// Cache load modifier for reading input elements
+  static constexpr cub::CacheLoadModifier LOAD_MODIFIER = cub::CacheLoadModifier::LOAD_DEFAULT;
+  constexpr static int ITEMS_PER_THREAD                 = NOMINAL_ITEMS_PER_THREAD_4B;
+  constexpr static int BLOCK_THREADS                    = NOMINAL_BLOCK_THREADS_4B;
+};
+
+template <int ItemsPerThread, int BlockSize>
+struct hub_t
+{
+  struct Policy : cub::ChainedPolicy<300, Policy, Policy>
+  {
+    constexpr static int ITEMS_PER_THREAD = ItemsPerThread;
+
+    using ReducePolicy = AgentReducePolicy<BlockSize, ItemsPerThread>;
+
+    // SingleTilePolicy
+    using SingleTilePolicy = ReducePolicy;
+
+    // SegmentedReducePolicy
+    using SegmentedReducePolicy = ReducePolicy;
+
+    // ReduceNondeterministicPolicy
+    using ReduceNondeterministicPolicy = ReducePolicy;
+  };
+
+  using MaxPolicy = Policy;
+};
+
+C2H_TEST("Nondeterministic Device reduce works with float and double on gpu",
+         "[reduce][nondeterministic]",
+         float_type_list)
+{
+  using type          = typename c2h::get<0, TestType>;
+  const int num_items = 1 << 20;
+  c2h::device_vector<type> d_input(num_items);
+  c2h::gen(C2H_SEED(2), d_input, static_cast<type>(-1000.0), static_cast<type>(1000.0));
+
+  c2h::device_vector<type> d_output(1);
+
+  const auto env = cuda::execution::require(cuda::execution::determinism::not_guaranteed);
+  cub::DeviceReduce::Reduce(d_input.begin(), d_output.begin(), num_items, cuda::std::plus<type>{}, type{}, env);
```

Comment:
Q: Is there a reason why this test cannot use the launch helper to also test CDP and CUDA graph invocation of this API?

See the use of `DECLARE_LAUNCH_WRAPPER` and `// %PARAM% TEST_LAUNCH lid 0:1:2` in other tests. Let me know if you need more background information on how this works.

Reply:
There is an existing test (`catch2_test_device_reduce_env.cu`) which uses launch wrappers to check that the requested determinism is respected. Having launch wrappers here would be redundant and would add compilation + execution time. We would also have to add a check for which kernels are being called (as we are currently doing in the other test I mentioned). I think we should only check reduce functionality here.

Reply:
Fine.